### PR TITLE
Automatically add newline to trace output

### PR DESCRIPTION
**Pull request by sim642**
It has always been annoying that every trace call must include a trailing newline in its format string. Otherwise trace output becomes unusable because multiple steps are printed on the same line.
We can just have the tracing module itself add the newline, just like how it adds the `%%%` prefix with indentation.

While adapting all trace calls, I found a couple in central parts of base analysis which weren't guarded by `if M.tracing then`.

### TODO
- [x] Update documentation

### CODE REVIEW

File: src/analyses/base.ml [hunk 1]
```diff
@@ -1935,7 +1935,7 @@ struct
     if M.tracing && exps <> [] then (
       let addrs = List.map (Tuple3.first) invalids' in
       let vs = List.map (Tuple3.third) invalids' in
-      M.tracel "invalidate" "Setting addresses [%a] to values [%a]\n" (d_list ", " AD.pretty) addrs (d_list ", " VD.pretty) vs
+      if M.tracing then M.tracel "invalidate" "Setting addresses [%a] to values [%a]" (d_list ", " AD.pretty) addrs (d_list ", " VD.pretty) vs
```
Comment (michael-schwarz) [inline src/analyses/base.ml:1938]:
```suggestion
      M.tracel "invalidate" "Setting addresses [%a] to values [%a]" (d_list ", " AD.pretty) addrs (d_list ", " VD.pretty) vs
```

This already is inside a `if M.tracing`.

File: src/analyses/loopTermination.ml [hunk 2]
```diff
@@ -74,7 +74,7 @@ struct
     | Queries.MustTermAllLoops ->
       let multithreaded = ctx.ask Queries.IsEverMultiThreaded in
       if multithreaded then (
-        M.warn ~category:Termination "The program might not terminate! (Multithreaded)\n";
+        M.warn ~category:Termination "The program might not terminate! (Multithreaded)";
```
Comment (michael-schwarz) [inline src/analyses/loopTermination.ml:77]: Does this PR  also make changes to warnings?

File: src/analyses/mCP.ml [hunk 3]
```diff
@@ -58,7 +58,7 @@ struct
       let y = find_id yn in
       if not (exists (fun (y',_) -> y=y') xs) then begin
         let xn = find_spec_name x in
-        Logs.error "Activated analysis '%s' depends on '%s' and '%s' is not activated.\n" xn yn yn;
+        Logs.error "Activated analysis '%s' depends on '%s' and '%s' is not activated." xn yn yn;
```
Comment (michael-schwarz) [inline src/analyses/mCP.ml:61]: Why the change to `Logs`?

File: src/analyses/memLeak.ml [hunk 4]
```diff
@@ -169,7 +169,7 @@ struct
       let allocated_and_unreachable_mem = D.diff allocated_mem reachable_mem in
       if not (D.is_empty allocated_and_unreachable_mem) then (
         set_mem_safety_flag InvalidMemTrack;
-        M.warn ~category:(Behavior (Undefined MemoryLeak)) ~tags:[CWE 401] "There is unreachable allocated heap memory at program exit. A memory leak might occur for the alloc vars %a\n" (Pretty.d_list ", " CilType.Varinfo.pretty) (D.elements allocated_and_unreachable_mem)
+        M.warn ~category:(Behavior (Undefined MemoryLeak)) ~tags:[CWE 401] "There is unreachable allocated heap memory at program exit. A memory leak might occur for the alloc vars %a" (Pretty.d_list ", " CilType.Varinfo.pretty) (D.elements allocated_and_unreachable_mem)
```
Comment (michael-schwarz) [inline src/analyses/memLeak.ml:172]: Why remove the `\n` here?

File: src/common/framework/cfgTools.ml [hunk 5]
```diff
@@ -148,7 +148,7 @@ let createCFG (file: file) =
      which do not otherwise appear in the control flow graph. *)
   let skippedByEdge = CfgEdgeH.create 113 in
 
-  if Messages.tracing then Messages.trace "cfg" "Starting to build the cfg.\n\n";
+  if Messages.tracing then Messages.trace "cfg" "Starting to build the cfg.";
```
Comment (michael-schwarz) [inline src/common/framework/cfgTools.ml:151]: Any reason we had `\n\n` here?

File: src/common/framework/cfgTools.ml [hunk 6]
```diff
@@ -163,7 +163,7 @@ let createCFG (file: file) =
     H.modify_def [] toNode (List.cons (edges,fromNode)) cfgB;
     H.modify_def [] fromNode (List.cons (edges,toNode)) cfgF;
     CfgEdgeH.replace skippedByEdge (fromNode, edges, toNode) skippedStatements;
-    if Messages.tracing then Messages.trace "cfg" "done\n\n"
+    if Messages.tracing then Messages.trace "cfg" "done"
```
Comment (michael-schwarz) [inline src/common/framework/cfgTools.ml:166]: Any reason we had `\n\n` here?

File: src/common/framework/cfgTools.ml [hunk 7]
```diff
@@ -468,7 +468,7 @@ let createCFG (file: file) =
           raise (Not_connect fd)
       | _ -> ()
     );
-  if Messages.tracing then Messages.trace "cfg" "CFG building finished.\n\n";
+  if Messages.tracing then Messages.trace "cfg" "CFG building finished.";
```
Comment (michael-schwarz) [inline src/common/framework/cfgTools.ml:471]: Any reason we had `\n\n` here?

File: src/framework/analyses.ml [hunk 8]
```diff
@@ -33,7 +33,7 @@ struct
   let getLocation (n,d) = Node.location n
 
   let pretty_trace () ((n,c) as x) =
-    if get_bool "dbg.trace.context" then dprintf "(%a, %a) on %a \n" Node.pretty_trace n LD.pretty c CilType.Location.pretty (getLocation x)
+    if get_bool "dbg.trace.context" then dprintf "(%a, %a) on %a" Node.pretty_trace n LD.pretty c CilType.Location.pretty (getLocation x)
```
Comment (michael-schwarz) [inline src/framework/analyses.ml:36]: correct?

File: src/framework/constraints.ml [hunk 9]
```diff
@@ -409,7 +409,7 @@ struct
       if ContextUtil.should_keep ~isAttr:GobContext ~keepOption:"ana.context.widen" ~keepAttr:"widen" ~removeAttr:"no-widen" f then (
         let v_old = M.find f.svar m in (* S.D.bot () if not found *)
         let v_new = S.D.widen v_old (S.D.join v_old v_cur) in
-        Messages.(if tracing && not (S.D.equal v_old v_new) then tracel "widen-context" "enter results in new context for function %s\n" f.svar.vname);
+        Messages.(if tracing && not (S.D.equal v_old v_new) then tracel "widen-context" "enter results in new context for function %s" f.svar.vname);
```
Comment (michael-schwarz) [inline src/framework/constraints.ml:412]: Why this unusual way to write it?

File: src/maingoblint.ml [hunk 10]
```diff
@@ -135,7 +135,7 @@ let check_arguments () =
     failwith "Option error"
   in
   let warn m = Logs.warn "%s" m in
-  if get_bool "allfuns" && not (get_bool "exp.earlyglobs") then (set_bool "exp.earlyglobs" true; warn "allfuns enables exp.earlyglobs.\n");
+  if get_bool "allfuns" && not (get_bool "exp.earlyglobs") then (set_bool "exp.earlyglobs" true; warn "allfuns enables exp.earlyglobs.");
```
Comment (michael-schwarz) [inline src/maingoblint.ml:138]: Was the extra newline on purpose?

File: src/solver/generic.ml [hunk 11]
```diff
@@ -63,21 +63,21 @@ struct
...
     incr SolverStats.evals;
     if (get_bool "dbg.solver-progress") then (incr stack_d; Logs.debug "%d" !stack_d)
 
   let update_var_event x o n =
     if tracing then increase x;
     if full_trace || ((not (Dom.is_bot o)) && Option.is_some !max_var && Var.equal (Option.get !max_var) x) then begin
-      if tracing then tracei "sol_max" "(%d) Update to %a.\n" !max_c Var.pretty_trace x;
-      if tracing then traceu "sol_max" "%a\n\n" Dom.pretty_diff (n, o)
+      if tracing then tracei "sol_max" "(%d) Update to %a" !max_c Var.pretty_trace x;
+      if tracing then traceu "sol_max" "%a" Dom.pretty_diff (n, o)
```
Comment (michael-schwarz) [inline src/solver/generic.ml:80]: `full_trace` and `trace` are unused inconsistently above and here. Should `full_trace` on its cause tracing?

File: src/solver/sLR.ml [hunk 12]
```diff
@@ -401,11 +401,11 @@ module Make0 =
...
           let restart_mode_x = h_find_default restart_mode x (2*GobConfig.get_int "solvers.slr4.restart_count") in
           let rstrt = use_box && (V.ver>3) && D.leq tmp old && restart_mode_x <> 0 in
-          if tracing then trace "sol" "Var: %a\n" S.Var.pretty_trace x ;
-          if tracing then trace "sol" "Contrib:%a\n" S.Dom.pretty tmp;
+          if tracing then trace "sol" "Var: %a" S.Var.pretty_trace x ;
+          if tracing then trace "sol" "Contrib:%a" S.Dom.pretty tmp;
           let tmp = if use_box then box old tmp else tmp in
           if not (D.eq tmp old) then begin
-            if tracing then trace "sol" "New Value:%a\n\n" S.Dom.pretty tmp;
+            if tracing then trace "sol" "New Value:%a" S.Dom.pretty tmp;
```
Comment (michael-schwarz) [inline src/solver/sLR.ml:408]: Double newline probably on purpose to separate potentially long printout of value from rest?

File: src/solver/sLRphased.ml [hunk 13]
```diff
@@ -101,28 +101,28 @@ module Make =
...
         in
-        if tracing then trace "sol" "Var: %a\n" S.Var.pretty_trace x ;
-        if tracing then trace "sol" "Contrib:%a\n" S.Dom.pretty val_new;
+        if tracing then trace "sol" "Var: %a" S.Var.pretty_trace x ;
+        if tracing then trace "sol" "Contrib:%a" S.Dom.pretty val_new;
         if S.Dom.equal old val_new then ()
         else begin
           update_var_event x old val_new;
-          if tracing then trace "sol" "New Value:%a\n\n" S.Dom.pretty val_new;
+          if tracing then trace "sol" "New Value:%a" S.Dom.pretty val_new;
```
Comment (michael-schwarz) [inline src/solver/sLRphased.ml:125]: `n\n` on purpose?

File: src/solver/sLRterm.ml [hunk 14]
```diff
@@ -171,12 +171,12 @@ module SLR3term =
...
         in
-        if tracing then trace "sol" "Var: %a\n" S.Var.pretty_trace x ;
-        if tracing then trace "sol" "Contrib:%a\n" S.Dom.pretty val_new;
+        if tracing then trace "sol" "Var: %a" S.Var.pretty_trace x ;
+        if tracing then trace "sol" "Contrib:%a" S.Dom.pretty val_new;
         if S.Dom.equal old val_new then ()
         else begin
           update_var_event x old val_new;
-          if tracing then trace "sol" "New Value:%a\n\n" S.Dom.pretty val_new;
+          if tracing then trace "sol" "New Value:%a" S.Dom.pretty val_new;
```
Comment (michael-schwarz) [inline src/solver/sLRterm.ml:179]: Double `\n\n`?

File: src/solver/topDown.ml [hunk 15]
```diff
@@ -59,22 +59,22 @@ module WP =
...
+          if tracing then trace "sol" "Contrib:%a" S.Dom.pretty tmp;
           let tmp = if is_side x then S.Dom.widen old (S.Dom.join old tmp) else if wpx then box old tmp else tmp in
           HM.remove called x;
           if not (S.Dom.equal old tmp) then (
-            if tracing then if is_side x then trace "sol2" "solve side: old = %a, tmp = %a, widen = %a\n" S.Dom.pretty old S.Dom.pretty tmp S.Dom.pretty (S.Dom.widen old (S.Dom.join old tmp));
+            if tracing then if is_side x then trace "sol2" "solve side: old = %a, tmp = %a, widen = %a" S.Dom.pretty old S.Dom.pretty tmp S.Dom.pretty (S.Dom.widen old (S.Dom.join old tmp));
             update_var_event x old tmp;
-            if tracing then trace "sol" "New Value:%a\n\n" S.Dom.pretty tmp;
-            if tracing then trace "sol2" "new value for %a (wpx: %b, is_side: %b) is %a. Old value was %a\n" S.Var.pretty_trace x (HM.mem rho x) (is_side x) S.Dom.pretty tmp S.Dom.pretty old;
+            if tracing then trace "sol" "New Value:%a" S.Dom.pretty tmp;
```
Comment (michael-schwarz) [inline src/solver/topDown.ml:69]: Any reason we had `\n\n` here?

File: src/solver/topDown_deprecated.ml [hunk 16]
```diff
@@ -56,21 +56,21 @@ module TD3 =
...
-          if tracing then trace "sol" "Contrib:%a\n" S.Dom.pretty tmp;
+          if tracing then trace "sol" "Var: %a" S.Var.pretty_trace x ;
+          if tracing then trace "sol" "Contrib:%a" S.Dom.pretty tmp;
           let tmp = if is_side x then S.Dom.widen old (S.Dom.join old tmp) else if wpx then box old tmp else tmp in
           HM.remove called x;
           if not (S.Dom.equal old tmp) then begin
             update_var_event x old tmp;
-            if tracing then trace "sol" "New Value:%a\n\n" S.Dom.pretty tmp;
-            if tracing then trace "sol2" "new value for %a (wpx: %b, is_side: %b) is %a. Old value was %a\n" S.Var.pretty_trace x wpx (is_side x) S.Dom.pretty tmp S.Dom.pretty old;
+            if tracing then trace "sol" "New Value:%a" S.Dom.pretty tmp;
```
Comment (michael-schwarz) [inline src/solver/topDown_deprecated.ml:65]: Any reason we had `\n\n` here?

File: src/solver/topDown_space_cache_term.ml [hunk 17]
```diff
@@ -29,34 +29,34 @@ module WP =
...
-          if tracing then trace "cache" "cache size %d for %a\n" (HM.length l) S.Var.pretty_trace x;
+          if tracing then trace "cache" "cache size %d for %a" (HM.length l) S.Var.pretty_trace x;
           cache_sizes := HM.length l :: !cache_sizes;
           if not (S.Dom.equal old tmp) then (
-            (* if tracing then if is_side x then trace "sol2" "solve side: old = %a, tmp = %a, widen = %a\n" S.Dom.pretty old S.Dom.pretty tmp S.Dom.pretty (S.Dom.widen old (S.Dom.join old tmp)); *)
+            (* if tracing then if is_side x then trace "sol2" "solve side: old = %a, tmp = %a, widen = %a" S.Dom.pretty old S.Dom.pretty tmp S.Dom.pretty (S.Dom.widen old (S.Dom.join old tmp)); *)
             update_var_event x old tmp;
-            if tracing then trace "sol" "New Value:%a\n\n" S.Dom.pretty tmp;
-            (* if tracing then trace "sol2" "new value for %a (wpx: %b, is_side: %b) is %a. Old value was %a\n" S.Var.pretty_trace x (HM.mem rho x) (is_side x) S.Dom.pretty tmp S.Dom.pretty old; *)
+            if tracing then trace "sol" "New Value:%a" S.Dom.pretty tmp;
```
Comment (michael-schwarz) [inline src/solver/topDown_space_cache_term.ml:58]: Any reason we had `\n\n` here?

File: src/solver/topDown_term.ml [hunk 18]
```diff
@@ -50,15 +50,15 @@ module WP =
...
+          if tracing then trace "sol" "Contrib:%a" S.Dom.pretty tmp;
           HM.remove called x;
           let tmp = if wpx then match phase with Widen -> S.Dom.widen old (S.Dom.join old tmp) | Narrow -> S.Dom.narrow old tmp else tmp in
           if not (S.Dom.equal old tmp) then (
-            (* if tracing then if is_side x then trace "sol2" "solve side: old = %a, tmp = %a, widen = %a\n" S.Dom.pretty old S.Dom.pretty tmp S.Dom.pretty (S.Dom.widen old (S.Dom.join old tmp)); *)
+            (* if tracing then if is_side x then trace "sol2" "solve side: old = %a, tmp = %a, widen = %a" S.Dom.pretty old S.Dom.pretty tmp S.Dom.pretty (S.Dom.widen old (S.Dom.join old tmp)); *)
             update_var_event x old tmp;
-            if tracing then trace "sol" "New Value:%a\n\n" S.Dom.pretty tmp;
-            (* if tracing then trace "sol2" "new value for %a (wpx: %b, is_side: %b) is %a. Old value was %a\n" S.Var.pretty_trace x (HM.mem rho x) (is_side x) S.Dom.pretty tmp S.Dom.pretty old; *)
+            if tracing then trace "sol" "New Value:%a" S.Dom.pretty tmp;
```
Comment (michael-schwarz) [inline src/solver/topDown_term.ml:60]: Any reason we had `\n\n` here?

File: src/witness/witnessConstraints.ml [hunk 19]
```diff
@@ -221,16 +221,16 @@ struct
...
-              if M.tracing then M.tracec "witness" "n = %a\n" Node.pretty_plain n;
-              if M.tracing then M.tracec "witness" "c = %a\n" Spec.C.pretty c;
-              if M.tracing then M.tracec "witness" "j = %a\n" Spec.D.pretty j;
+              if M.tracing then M.tracec "witness" "n = %a" Node.pretty_plain n;
+              if M.tracing then M.tracec "witness" "c = %a" Spec.C.pretty c;
+              if M.tracing then M.tracec "witness" "j = %a" Spec.D.pretty j;
               f (I.to_int x) (n, Obj.repr c, I.to_int j) e
             ) r;
-          if M.tracing then M.traceu "witness" "\n"
+          if M.tracing then M.traceu "witness" ""
```
Comment (michael-schwarz) [inline src/witness/witnessConstraints.ml:233]: Maybe a comment so this isn't deleted?

File: src/analyses/loopTermination.ml (hunk 2 above)
Comment (sim642) [reply src/analyses/loopTermination.ml:77]: Only the handful of instances that accidentally end with `\n` and thus introduce random empty lines between warnings. These are typos from having had to keep in mind the two different requirements for newlines for messages/logs and tracing. By normalizing all three, none of them should end with a `\n`.

File: src/framework/analyses.ml (hunk 8 above)
Comment (sim642) [reply src/framework/analyses.ml:36]: Yes, because an individual `pretty`, which is used between other stuff in larger format strings, has no need to unconditionally put a newline after itself. That makes it impossible to include this `pretty` in a single-line output.

File: src/framework/constraints.ml (hunk 9 above)
Comment (sim642) [reply src/framework/constraints.ml:412]: I guess to avoid writing `Messages.` twice. This PR is not to change that.

**Comment by sim642** [comment]
As for all the `\n\n`-s, I see no reason to have tons of double empty lines in already-way-too-long tracing output. Large printed abstract values are printed indented after their `%%%` line and end where the next `%%%` appears for the following tracing at the beginning of the line. So they're already delimited by the `%%%` lines and the two empty lines just mean more scrolling.

Also, empty lines can be annoying/misleading because often tracing output has values with massive (500+) levels of indentation, so it's not obvious whether it's empty or there's something off-screen far right, without carefully scrolling horizontally. The `%%%` are at the beginnings of separating lines, so you can already tell where one large value ends just by the beginning of the line.

File: src/solver/generic.ml (hunk 11 above)
Comment (sim642) [reply src/solver/generic.ml:80]: I don't think so, because in different places `full_trace` is used both positively and negatively to do `trace` calls, so it's intended to be some kind of switch.
I made a few calls also conditional on `tracing`, so that controls everything like it should be. `full_trace` just changes what's being traced (if anything).

**Review by michael-schwarz — APPROVED** [review]
